cpp/unreal_plugins/SpEngine/Source/SpEngine/EngineService.h

```diff
@@ -0,0 +1,175 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#pragma once
+
+#include <atomic>
+#include <concepts> // std::same_as
+#include <future>   // std::promise, std::future
+#include <string>
+
+#include <Delegates/IDelegateInstance.h> // FDelegateHandle
+#include <Misc/CoreDelegates.h>
+
+#include "SpCore/Assert.h"
+#include "SpCore/Log.h"
+#include "SpEngine/WorkQueue.h"
+
+template <typename TBasicEntryPointBinder>
```

Comment:
I like this naming convention better.

```
template <typename TEntryPointBinder>
concept CEntryPointBinder = requires(TEntryPointBinder entry_point_binder) {
    { entry_point_binder.bind("", []() -> void {}) } -> std::same_as<void>;
};

template <typename TUnrealEntryPointBinder>
concept CUnrealEntryPointBinder = requires(TUnrealEntryPointBinder unreal_entry_point_binder) {
    { unreal_entry_point_binder.bindFuncUnreal("", "", []() -> void {}) } -> std::same_as<void>;
    { unreal_entry_point_binder.bindFuncNoUnreal("", "", []() -> void {}) } -> std::same_as<void>;
};
```

If a service is calling `EngineService::bindFuncUnreal(...)` or `EngineService::bindFuncNoUnreal(...)`, they shouldn't need to concern themselves with the implementation details of `EngineService`. The fact that `EngineService` sometimes wraps functions that get passed in is an implementation detail, and a user that is unfamiliar with how `EngineService` is implemented probably wouldn't understand at first glance which function is correct for their use case.

_"Wait, what does wrapped mean here? Do I want to use `bind_func` or `bind_func_wrapped`? I'm not sure."_

A caller should only be responsible for specifying if their function will interact with the Unreal Engine or not, which is why I prefer the names `bindFuncUnreal` and `bindFuncNoUnreal`.

Reply:
That makes more sense. I also did not like `bind_func` and `bind_func_wrapped`.